File: Behavioral/DynProg/DynProg.py
import numpy as np
from scipy.optimize import minimize
import math
import matplotlib.pyplot as plt

# Initialize the parameters of the problem.

alpha = .4
beta = .9
gamma = .5
delta = .1

k0 = .6

# It actually takes a little bit if this is made much larger.
I = 5

# The steady state comes from the point where capital and consumption
# are unchanging, so:
# k = k^alpha + (1-delta)k - c
# Solve to reach: c =  k^alpha -  delta k
# Applying the golden rule, AKA max consumption so del c del k = 0
# alpha k^(alpha-1) - delta = 0
# k = (delta/alpha)^(1/(alpha-1))

#kStar = (delta / alpha)**(1 / (alpha - 1))
kStar = (((1 / beta) + delta - 1) / alpha)**(1. / (alpha - 1.))
cStar = kStar**alpha - delta * kStar

Epsilon = .00001

# Now we need to map our points from k0 to kStar + (kStar - k0) to -1,1
a = k0
b = 2 * kStar - k0


def MoveToChebyDomain(x):
    # Go from (a,b) to (-1,1)
    return 2 * (x - a) / (b - a) - 1


def MoveToProblemDomain(x):
    return a + (x + 1) * ((b - a) / 2)

# Following the Notation of Paarsch and Golyeav


def LSFit(y, bT, M):
    a = np.zeros(M)
    for m in range(0, M):
        a[m] = sum(y * bT[m, :]) / sum(bT[m, :] * bT[m, :])

    return a


# This again follows the work in "Effective Computing in Quantitative Research"
def BuildChebyCoeff(Xk, v, k):
    M = k - 1
    # Initialize the matrix, setting the second row
    bT = np.ones((M, k))
    bT[1, :] = Xk
    # Recursively fill the matrix using the Chebyshev polynomials
    for m in range(2, M):
        bT[m, :] = 2. * Xk * bT[m - 1, :] - bT[m - 2, :]
    # Fit the coefficients using Least Squares ala Paarsch/Golyeav
    aHat = LSFit(v, bT, M)
    return aHat

# This builds the chebyshev polynomials recursively


def Chebyshev(index, xVal):
    # Since T_0 = 1 and T_1 = x
    # And T_n = 2xT_n-1 - T_n-2
    if(index == 0):
        return 1

    if(index == 1):
        return xVal

    return(2 * xVal * Chebyshev(index - 1, xVal) - Chebyshev(index - 2, xVal))


# This evaluates Chebyshev polynomials for some coefficients and some x value - Having issues vectoring this.
def EvalChebyCoeff(A, xVal):
    # Evaluate a ChebyChev Polynomial
    # with coefficients a
    # at x and return its value.
    value = 0

    for i in range(len(A)):
        value += A[i] * (Chebyshev(i, xVal))
    return value


k = 2 * I + 1

Xk = np.zeros(k)
for i in range(1, k + 1):
    Xk[i - 1] = -math.cos(math.pi * (2 * i - 1) / (2 * k))

# So I am just going to initialize this with the constant steady state level
# of consumption, and we will see what happens from there
v = np.ones(k) * cStar * (1 / (1 - beta))


A = BuildChebyCoeff(Xk, v, k)


def OptConsume(c, k, a):
    return -1 * (c[0] ** gamma + beta * EvalChebyCoeff(a, MoveToChebyDomain(k**alpha - c[0])))


# Hopefully this does the trick
prevV = np.zeros(k)
cVec = np.ones(k) * cStar
iterCounter = 0
# Whats a reasonable limit on the number of iterations? I was getting ~ 150
while(np.linalg.norm(prevV - v) > Epsilon):
    prevV = np.copy(v)
    for i in range(k):
        curK = MoveToProblemDomain(Xk[i])
        cLow = 0
        cHigh = curK ** alpha + (1 - delta) * curK
        #curK = ((cLow + cHigh) / 2)
        # Choose the C in this range that maximizes c^gamma + beta*V( k^alpha - c )
        # Since we don't know V, all we have is our best guess formed by v.
        cBest = minimize(OptConsume, [(cLow + cHigh) / 2],
                         (curK, A), 'TNC', bounds=[(cLow, cHigh)])
        v[i] = -cBest.fun
        cVec[i] = cBest.x[0]
    # Now we have a new V. So we can estimate a better a.
    A = BuildChebyCoeff(Xk, v, k)
    iterCounter += 1

C = BuildChebyCoeff(Xk, cVec, k)
#C = np.polynomial.chebyshev.chebfit(Xk, cVec, k)

# Supposedly we have a good fit for V now.
# print(v)
# print(A)
print(iterCounter)

# This is plotting nonsense.
t = np.arange(0., MoveToProblemDomain(max(Xk)), .05)

#s = np.polynomial.chebyshev.chebval(t, C)
s = EvalChebyCoeff(C, MoveToChebyDomain(t))
#u = np.polynomial.chebyshev.chebval(t, A)
fig, ax = plt.subplots()
ax.plot(t, s)
#ax.plot(t, u)

ax.grid()

fig.savefig("test.png")
plt.show()

File: Behavioral/DynProg/DynProg.jl
using Optim
using Plots
# Initialize the parameters of the problem.

alpha = .4
beta = .9
gamma = .5
delta = .1

k0 = .6

# It actually takes a little bit if this is made much larger.
I = 10

# These values have been changed to the steady states given by Paarsch in Email

kStar = (((1/beta)+delta-1) / alpha)^(1./(alpha-1.))

cStar = kStar^alpha-delta*kStar

Epsilon = .00001

# Now we need to map our points from k0 to kStar + (kStar - k0) to -1,1
a = k0
b = 2 * kStar - k0

#Since the Domain of the ChebyShev polynomials is -1,1, we want to move our work into this domain and out
function MoveToChebyDomain( x::Float64 )
    # Go from (a,b) to (-1,1)
    return 2 * (x - a) / (b - a) - 1
end

function MoveToProblemDomain( x::Float64 )
    return a + (x + 1) * ((b - a) / 2)
end

#Following the Notation of Paarsch and Golyeav
function LSFit( y::Vector{Float64}, bT::Matrix{Float64}, M::Integer )
    # a = zeros( M )
    # for m in 1:M
    #     a[m] = sum( y .* bT[m,:])/sum(bT[m,:] .* bT[m,:])
    # end
    return (bT*bT') \ (bT*y)
end

#This again follows the work in "Effective Computing in Quantitative Research"
function BuildChebyCoeff( Xk::Vector{Float64}, v::Vector{Float64}, k::Integer)
    M = k-1
    #Initialize the matrix, setting the second row
    bT = ones( M, k)
    bT[2,:] = Xk
    #Recursively fill the matrix using the Chebyshev polynomials
    #The dots in front of operators are vectorizing the operation
    for m in 3:M
        bT[m,:] = 2.*Xk .* bT[m-1,:]  .- bT[m-2,:]
    end
    #Fit the coefficients using Least Squares ala Paarsch/Golyeav
    aHat = LSFit( v, bT, M)
    return aHat
end

#This builds the chebyshev polynomials recursively
function Chebyshev( index::Integer, xVal::Float64 )
    #Since T_0 = 1 and T_1 = x
    #And T_n = 2xT_n-1 - T_n-2
    if( index == 1 )
        return 1
    end
    if( index == 2)
        return xVal
    end
    return( 2*xVal*Chebyshev(index-1,xVal) - Chebyshev(index-2,xVal))
end

#This evaluates Chebyshev polynomials for some coefficients and some x value - Having issues vectoring this.
function EvalChebyCoeff( A::Vector{Float64}, xVal::Float64 )
    #Evaluate a ChebyChev Polynomial
    #with coefficients a
    #at x and return its value.
    value = 0

    for i in 1:length(A)
        value += A[i]*(Chebyshev(i,xVal))
    end
    return value
end
    
#Find the Chebyshev nodes for k
k = 2 * I + 1
Xk = zeros( k )
for i in 1:k
    Xk[i] = -cos( π*(2*i - 1)/(2*k))
end

#This is the chebyshev nodes located in The Capital Space [a,b]
x = a .+ ((Xk+1).*((b-a)/2))

#Here is our initial guess at valuation - We consume at the steady state of capital
v = ones(k)*cStar*(1/(1-beta))

#Fit the chebyshev polynomial to this valuation
A = BuildChebyCoeff( Xk, v, k)

#Set everything up for the loop
prevV = zeros(k)
cVec = ones(k)*cStar
iterCounter = 0

#While our function is sufficiently different at the points we are "training" it
while( norm(prevV - v) > Epsilon )
    prevV = copy(v)
    for i in 1:k
        #Get the capital level, and the max and min consumption
        curK = x[i]
        cLow = 0
        cHigh = curK^alpha + (1-delta)*curK
        
        #Consumption for some c is given by:
        # c^gamma + beta*v(k^alpha - c)
        #We want to find the c that maximizes this.
        result = optimize(
            (x->-(x^gamma + beta*EvalChebyCoeff(A, MoveToChebyDomain(curK^alpha - x) ))),
            cLow, cHigh)#; alpha, gamma, beta, curK )
        #There is a negative here because we minimized the negative of consumption
        v[i] = -Optim.minimum(result)
        #This is only used on the final loop, should be better way of handing this.
        cVec[i] = Optim.minimizer(result)
    end
    #Now we have a better guess of v, so we can build a better polynomial
    A = BuildChebyCoeff( Xk, v, k)
    iterCounter += 1
end

#Now build our consumption polynomial
C = BuildChebyCoeff( Xk, cVec, k)
#How long did it take us to get there
print( iterCounter )


#This only looks so bad because I haven't managed to vectorize EvalChebyCoeff properly.
# x = linspace( a,b)
# y = zeros( length( x ))
# for xi in x
#     y = EvalChebyCoeff( C, MoveToChebyDomain(xi) )
# end

# #Plot our consumption and label it accordingly
# pyplot()
# plot( x,EvalChebyCoeff( C, MoveToChebyDomain(x) ),title="Consumption", label="",lw=3)
# xlabel!( "Current Capital")
# ylabel!( "Consumption")
# savefig("consumptionPath.pdf") 
